securedrop/tests/test_i18n.py

```diff
@@ -0,0 +1,148 @@
+
+        with app.test_client() as c:
+            c.get('/')
+            assert session.get('locale') is None
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=fr_FR', headers=Headers([('Accept-Language', 'en_US')]))
+            assert session.get('locale') == 'fr_FR'
+            assert translated_fr == gettext(not_translated)
+
+            c.get('/', headers=Headers([('Accept-Language', 'en_US')]))
+            assert session.get('locale') == 'fr_FR'
+            assert translated_fr == gettext(not_translated)
+
+            c.get('/?l=')
+            assert session.get('locale') is None
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=en_US', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') == 'en_US'
+            assert not_translated == gettext(not_translated)
+
+            c.get('/', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') == 'en_US'
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') is None
+            assert translated_fr == gettext(not_translated)
+
+            c.get('/')
+            assert session.get('locale') is None
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=YY_ZZ')
+            assert session.get('locale') is None
+            assert not_translated == gettext(not_translated)
+
+    def test_i18n(self):
+        sources = [
+            'tests/i18n/code.py',
+            'tests/i18n/template.html',
+        ]
+        kwargs = {
+            'translations_dir': config.TEMP_DIR,
+            'mapping': 'tests/i18n/babel.cfg',
+            'source': sources,
+            'extract_update': True,
+            'compile': True,
+            'verbose': logging.DEBUG,
+            'version': version.__version__,
+        }
+        args = argparse.Namespace(**kwargs)
+        manage.setup_verbosity(args)
+        manage.translate(args)
+
+        manage.sh("""
+        pybabel init -i {d}/messages.pot -d {d} -l en_US
+        pybabel init -i {d}/messages.pot -d {d} -l fr_FR
+        sed -i -e '/code hello i18n/,+1s/msgstr ""/msgstr "code bonjour"/' \
+              {d}/fr_FR/LC_MESSAGES/messages.po
+        pybabel init -i {d}/messages.pot -d {d} -l zh_Hans_CN
+        sed -i -e '/code hello i18n/,+1s/msgstr ""/msgstr "code chinese"/' \
+              {d}/zh_Hans_CN/LC_MESSAGES/messages.po
+        """.format(d=config.TEMP_DIR))
+
+        manage.translate(args)
+
+        for app in (journalist.app, source.app):
+            app.config['BABEL_TRANSLATION_DIRECTORIES'] = config.TEMP_DIR
+            i18n.setup_app(app)
+            self.verify_i18n(app)
+
+    @classmethod
+    def teardown_class(cls):
+        reload(journalist)
+        reload(source)
```

Comment:
Thought: we could have a test to guard against forgetting that `DEFAULT_LOCALE` may not be set in `config.py`, something along the lines of:

```
def test_verify_default_locale_en_us_if_not_defined_in_config(self):
        DEFAULT_LOCALE = config.DEFAULT_LOCALE
        try:
            config.DEFAULT_LOCALE = None
            with app.test_client() as c:
                c.get('/')
                assert session.get('locale') == 'en_US'
                assert not_translated == gettext(not_translated)
        finally:
            config.DEFAULT_LOCALE = DEFAULT_LOCALE
```

Note: I have not tested this and it may need some modification, just jotting down the general idea

Reply:
Good idea ! To be perfectly honest I thought about adding a similar test and conveniently forgot about it ... Your scaffold was useful and adapted. Repushed !